File: cpp/strong_password.cpp
#include "bits/stdc++.h"
using namespace std;

const int N = 200005;

int get(char c) {
   if(c >= '0' && c <= '9') return 0;
   if(c >= 'a' && c <= 'z') return 1;
   if(c >= 'A' && c <= 'Z') return 2;
   return 3;
}

int minimumNumber(int n, string password) {
    // Return the minimum number of characters to make the password strong
    int num = 1, spec = 1, low=1, cap=1, rest = 2;
    string special_characters = "!@#$%^&*()-+";
    for(int i =0; i < password.size() ; i++) {
        if(num!=0 && password[i] >='0' && password[i]<='9')
            num--;
        else if(low!=0 && password[i]>='a' && password[i]<='z') 
            low--;
        
        else if(cap!= 0 && password[i]>='A' && password[i]<='Z')
            cap--;       
        
        else if(spec!=0 && special_characters.find(password[i]) != special_characters.npos) 
            spec--;
        
        else if(rest !=0) 
            rest--;
        
        else if(num==cap==spec==rest==0)    
            break;
    }
    return num+spec+low+cap+rest;
}

int main() {
   int n;
   string s;

   cin >> n;
   cin >> s;

   int ans = 0;

   for(int i = 0; i < n; i++) {
      ans |= (1 << get(s[i]));
   }

   cout << max(6 - n, 4 - __builtin_popcount(ans));
   return 0;
}

File: strong_password.py
#!/bin/python3
# 03/05/2022
# Link: https://www.hackerrank.com/challenges/three-month-preparation-kit-strong-password/problem

"""
Problem:

Given a password, find the minimum number of characters to add to make it strong.

Solution:

The answer is always `max(6-n, 4-d)` where `n` is string length and `d` is the number of different type of 
characters that are already present in the input password.

"""

import math
import os
import random
import re
import sys

#
# Complete the 'minimumNumber' function below.
#
# The function is expected to return an INTEGER.
# The function accepts following parameters:
#  1. INTEGER n
#  2. STRING password
#

# using regex Method #2
def minimumNumber1(n, password):
    MIN_LENGTH = 6
    charsToChange = 0
    
    if not re.search('[0-9]', password): charsToChange += 1
    if not re.search('[A-Z]', password): charsToChange += 1
    if not re.search('[a-z]', password): charsToChange += 1 
    if not re.search(r'[!@#$%^&*()\-+]', password): charsToChange += 1
    
    return max(MIN_LENGTH - len(password), charsToChange) if n < 6 else charsToChange

# Initial Method #1
def minimumNumber(n, password):
    # Return the minimum number of characters to make the password strong
        
    numbers = "0123456789"
    lower_case = "abcdefghijklmnopqrstuvwxyz"
    upper_case = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    special_characters = "!@#$%^&*()-+"
    
    pas = set(password)
    
    if n < 6:
        min_chars_req = 0
        nums = 0
        lo = 0
        hi = 0
        sp = 0
        for char in pas:
            if char in numbers:
                nums += 1
            elif char in lower_case:
                lo += 1
            elif char in upper_case:
                hi += 1
            elif char in special_characters:
                sp += 1
                
        if nums == 0:
            min_chars_req += 1
            
        if lo == 0:
            min_chars_req += 1
            
        if hi == 0:
            min_chars_req += 1
            
        if sp == 0:
            min_chars_req += 1
            
        if n + min_chars_req < 6:
            min_chars_req = 6 - n
            
    else:
        min_chars_req = 0
        nums = 0
        lo = 0
        hi = 0
        sp = 0
        for char in pas:
            if char in numbers:
                nums += 1
            elif char in lower_case:
                lo += 1
            elif char in upper_case:
                hi += 1
            elif char in special_characters:
                sp += 1
                
        if nums == 0:
            min_chars_req += 1
            
        if lo == 0:
            min_chars_req += 1
            
        if hi == 0:
            min_chars_req += 1
            
        if sp == 0:
            min_chars_req += 1
            
    return min_chars_req

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input().strip())

    password = input()

    answer = minimumNumber(n, password)

    fptr.write(str(answer) + '\n')

    fptr.close()
